liquibase-core/src/main/java/liquibase/command/core/TagCommandStep.java

```diff
@@ -19,34 +31,53 @@ public class TagCommandStep extends AbstractCliWrapperCommandStep {
 
     static {
         CommandBuilder builder = new CommandBuilder(COMMAND_NAME);
-        URL_ARG = builder.argument(CommonArgumentNames.URL, String.class).required()
-                .description("The JDBC database connection URL").build();
+        TAG_ARG = builder.argument("tag", String.class).required()
+                .description("Tag to add to the database changelog table").build();
+        URL_ARG = builder.argument(CommonArgumentNames.URL, String.class).description("The JDBC database connection URL").build();
         DEFAULT_SCHEMA_NAME_ARG = builder.argument("defaultSchemaName", String.class)
                 .description("The default schema name to use for the database connection").build();
         DEFAULT_CATALOG_NAME_ARG = builder.argument("defaultCatalogName", String.class)
                 .description("The default catalog name to use for the database connection").build();
         DRIVER_ARG = builder.argument("driver", String.class)
                 .description("The JDBC driver class").build();
         DRIVER_PROPERTIES_FILE_ARG = builder.argument("driverPropertiesFile", String.class)
-               .description("The JDBC driver properties file").build();
+                .description("The JDBC driver properties file").build();
         USERNAME_ARG = builder.argument(CommonArgumentNames.USERNAME, String.class)
                 .description("Username to use to connect to the database").build();
         PASSWORD_ARG = builder.argument(CommonArgumentNames.PASSWORD, String.class)
                 .description("Password to use to connect to the database")
                 .setValueObfuscator(ConfigurationValueObfuscator.STANDARD)
                 .build();
-        TAG_ARG = builder.argument("tag", String.class).required()
-                .description("Tag to add to the database changelog table").build();
+        DATABASE_ARG = builder.argument("database", Database.class).hidden().build();
     }
 
+
     @Override
-    public String[][] defineCommandNames() {
-        return new String[][] { COMMAND_NAME };
+    public void run(CommandResultsBuilder resultsBuilder) throws Exception {
+        CommandScope commandScope = resultsBuilder.getCommandScope();
+        this.setOrCreateDatabase(commandScope, DATABASE_ARG);
+        LockService lockService = LockServiceFactory.getInstance().getLockService(getDatabase());
+        lockService.waitForLock();
+
+        try {
+            ChangeLogHistoryServiceFactory.getInstance().getChangeLogService(getDatabase()).generateDeploymentId();
+            checkLiquibaseTables(false, null, new Contexts(), new LabelExpression());
+            getDatabase().tag(commandScope.getArgumentValue(TagCommandStep.TAG_ARG));
```

Comment:
The "tag" function on Database is just a call to `ChangeLogHistoryServiceFactory`. As we're consolidating the logic, now is probably a good time to mark that Database.tag method as deprecated and move the ChangeLogHistoryServiceFactory call to here instead.

Maybe even move the Database.tag() call the interface as a default method so subclasses that don't use AbstractJdbcDatabase don't have to worry about it anymore?

Could probalby do the same deprecation/default-method'ing with Database.doesTagExist() while you're at it too?

As we're refactoring these commands, it's not "automatically move all logic in here", like it we really did want tag() to be on Database we should have called out to there. And we're not moving the logic from the HistoryService into there. But still always be looking at what is in the called methods and be thinking about whether that logic should be in the command or belongs outside it.